v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/templates/KafkaToBigQueryFlex.java

```diff
@@ -0,0 +1,451 @@
+          .apply("Flatten", Flatten.pCollections())
+          .apply(
+              "PrintInsertionFailedRecords",
+              ParDo.of(new ThrowErrorFn<KV<String, String>, String>()));
+    }
+
+    if (options.getOutputDeadletterTable() != null) {
+      /*
+       * Step #5: Insert records that failed BigQuery inserts into a deadletter table.
+       */
+      failedInserts.apply(
+          "WriteInsertionFailedRecords",
+          ErrorConverters.WriteStringMessageErrors.newBuilder()
+              .setErrorRecordsTable(
+                  ObjectUtils.firstNonNull(
+                      options.getOutputDeadletterTable(),
+                      options.getOutputTableSpec() + DEFAULT_DEADLETTER_TABLE_SUFFIX))
+              .setErrorRecordsTableSchema(SchemaUtils.DEADLETTER_SCHEMA)
+              .build());
+    } else {
+      failedInserts.apply(
+          "PrintInsertionFailedRecords", ParDo.of(new ThrowErrorFn<String, String>()));
+    }
+
+    return pipeline.run();
+  }
+
+  /**
+   * Method to wrap a {@link BigQueryInsertError} into a {@link FailsafeElement}.
+   *
+   * @param insertError BigQueryInsert error.
+   * @return FailsafeElement object.
+   */
+  protected static FailsafeElement<String, String> wrapBigQueryInsertError(
+      BigQueryInsertError insertError) {
+
+    FailsafeElement<String, String> failsafeElement;
+    try {
+
+      failsafeElement =
+          FailsafeElement.of(
+              insertError.getRow().toPrettyString(), insertError.getRow().toPrettyString());
+      failsafeElement.setErrorMessage(insertError.getError().toPrettyString());
+
+    } catch (IOException e) {
+      LOG.error("Failed to wrap BigQuery insert error.");
+      throw new RuntimeException(e);
+    }
+    return failsafeElement;
+  }
+
+  /**
+   * Method to create Kafka Client Authentication Config with the given username and password.
+   *
+   * @param username Kafka username.
+   * @param password Kafka password.
+   * @return ImmutableMap object.
+   */
+  static ImmutableMap<String, Object> setClientAuthConfig(String username, String password) {
+    ImmutableMap.Builder<String, Object> properties = ImmutableMap.builder();
+    properties.put(SaslConfigs.SASL_MECHANISM, "PLAIN");
+    properties.put(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG, "SASL_SSL");
+    properties.put(
+        SaslConfigs.SASL_JAAS_CONFIG,
+        "org.apache.kafka.common.security.plain.PlainLoginModule required"
+            + " username=\'"
+            + username
+            + "\'"
+            + " password=\'"
+            + password
+            + "\';");
+    return properties.buildOrThrow();
+  }
+
+  static class ThrowErrorFn<T, W> extends DoFn<FailsafeElement<T, W>, FailsafeElement<T, W>> {
+
+    @ProcessElement
+    public void processElement(ProcessContext context) {
+      FailsafeElement<T, W> element = context.element();
+      LOG.error(element.toString() + element.getErrorMessage() + element.getStacktrace());
```

Comment:
OK to keep for now, but we can't have this in the production version of the template (can't log every error, this may overload the Cloud Logging API if there're too many errors). Maybe at least add a TODO to fix this later?

Reply:
Added a TODO